File: 101/04 二分查找/01_69.js
/**
 * @param {number} x
 * @return {number}
 */
var mySqrt = function (x) {
  let l = 0;
  let r = x;
  let mid = -1;
  let sqrt = -1;
  while (l <= r) {
    mid = l + Math.floor((r - l) / 2);
    if (BigInt(mid) * BigInt(mid) <= x) {
      sqrt = mid;
      l = mid + 1;
    } else {
      r = mid - 1;
    }
  }
  return sqrt;
};

File: 101/04 二分查找/01_69.cc
class Solution {
  public:
    int mySqrt(int x) {
      int l = 0;
      int r = x;
      int mid = -1;
      int sqrt = -1;
      while (l <= r) {
        mid = l + (r - l) / 2;
        if ((long long)mid * mid <= x) {
          sqrt = mid;
          l = mid + 1;
        } else {
          r = mid - 1;
        }
      }
      return sqrt;
    }
};
